### PR TITLE
PS-6864: Merge fb-prod201902 (5.7)

### PR DESCRIPTION
**Pull request by inikep**
1. Enable RFR at https://github.com/percona/percona-server/commit/2c85ced009da8d77228328ce1fd009ce5168cc7d

2. Port previously skipped blind deletes at https://github.com/percona/percona-server/commit/708d76d7cc7be36c5ad7c16fefd9d9ec1f970a25

3. Update rocksdb submodule to [51a90415e2203b063630277d95f92d8b95e09ed6 Add load statements to rocksdb TARGETS files](https://github.com/facebook/rocksdb/commit/51a90415e2203b063630277d95f92d8b95e09ed6) (v5.8-1323-g51a90415e)

4. Cherry-pick the following commits from `fb-prod201902`:
- [3ed8b062eaacb2a8dd9616bbf4d3f267af7ba91e Fix error message in MyRocks](https://github.com/facebook/mysql-5.6/commit/3ed8b062eaacb2a8dd9616bbf4d3f267af7ba91e)
- [dfbb8ffd7b0d6f1c8f3cb211edcc446ced5badee Skip snapshot validation on Read Committed (#920)](https://github.com/facebook/mysql-5.6/commit/dfbb8ffd7b0d6f1c8f3cb211edcc446ced5badee)
- [f55c956b4d8266bd6e4c8deabc2f69df34b6a3d4 Add Rdb_buf_writer to wrap rdb_netbuf_store_* on a static buffer and add more length validation. Also fix a circulr dependency between rdb_buff.h and ha_rocksdb.h by introducing rdb_global.h](https://github.com/facebook/mysql-5.6/commit/f55c956b4d8266bd6e4c8deabc2f69df34b6a3d4)
- [22ad6ef040758bad44ac3137cd14840ec2ceaebb Enable more debugging in rocksdb.bloomfilter3 to track down server hang](https://github.com/facebook/mysql-5.6/commit/22ad6ef040758bad44ac3137cd14840ec2ceaebb)
- [bb92e00f199c624b86fbe891ca0e4394837b48de Compaction filter can't find index info occasionally](https://github.com/facebook/mysql-5.6/commit/bb92e00f199c624b86fbe891ca0e4394837b48de)
- [dfa36bd3558e5d1786131af2925bf0091b005c69 small perf improvement when extracting TTL duration value](https://github.com/facebook/mysql-5.6/commit/dfa36bd3558e5d1786131af2925bf0091b005c69)
- [e702526d0f67daaaefd5501a1768505c7cb6e9df more rdb_buf_writer refactoring and fix a MySqlWhiteSpace linter bug](https://github.com/facebook/mysql-5.6/commit/e702526d0f67daaaefd5501a1768505c7cb6e9df)
- [3ed3a39d92514954628eb68bee1a24863c23a90b Issue-895 : Assertion `!options.use_direct_reads' failed. (#897)](https://github.com/facebook/mysql-5.6/commit/3ed3a39d92514954628eb68bee1a24863c23a90b)
- [ebd302a7efa8499eaca9281d1384998fe1cc93d6 Perf: Remove extra rdb_normalize_tablename calls](https://github.com/facebook/mysql-5.6/commit/ebd302a7efa8499eaca9281d1384998fe1cc93d6)
- [3115eacfce5551281497fb35bfd2f0102ffecda8 Fix test with non-deterministic snapshot count](https://github.com/facebook/mysql-5.6/commit/3115eacfce5551281497fb35bfd2f0102ffecda8)
- [0a17561c58711b28598fe22917657d1ae9a96739 Fix bulk-load test with non-deterministic snapshot count](https://github.com/facebook/mysql-5.6/commit/0a17561c58711b28598fe22917657d1ae9a96739)
- [f21dbccd354e54d6fb5621075c4cc367885860c5 Remove unused RDB_SST_INFO_USE_THREAD feature in bulk loading. It is not being used/tested and there is no point maintaining this going forward](https://github.com/facebook/mysql-5.6/commit/f21dbccd354e54d6fb5621075c4cc367885860c5)
- [eb28cee4551de8172f491e308a08386ad641b51d Add more comments for read/write code path](https://github.com/facebook/mysql-5.6/commit/eb28cee4551de8172f491e308a08386ad641b51d)
- [ffd481e6a8b3589a06a9cff9eb9ea3515fb23cc4 Remove unused parameters](https://github.com/facebook/mysql-5.6/commit/ffd481e6a8b3589a06a9cff9eb9ea3515fb23cc4)
- [a1bafccee7e912b509df98e9efc66487a30d54b4 Skip calculate matching length for key = m_index_number case](https://github.com/facebook/mysql-5.6/commit/a1bafccee7e912b509df98e9efc66487a30d54b4)
- [8cc282bc1a294237aa614eae60d057f62c625bf1 Use recreate for truncate partition](https://github.com/facebook/mysql-5.6/commit/8cc282bc1a294237aa614eae60d057f62c625bf1)
- [fb8be29447330ae950376259771ec6d252ae91e8 Move explicit snapshot related global vars into Rdb_explicit_snapshot.](https://github.com/facebook/mysql-5.6/commit/fb8be29447330ae950376259771ec6d252ae91e8)
- [b9d3bfbfa3cf1fc4227b48fd9705b511cc2409ee Perf: Use emplace_back instead push_back for Decoders](https://github.com/facebook/mysql-5.6/commit/b9d3bfbfa3cf1fc4227b48fd9705b511cc2409ee)
- [269d05b0beb761d9f9ba057f12b62b0077a3c4a2 Enable atomic bulk loading](https://github.com/facebook/mysql-5.6/commit/269d05b0beb761d9f9ba057f12b62b0077a3c4a2)
- [7ddbc3c6e127cac00e40b035157b41e19da85638 Remove friend class in Rdb_key_def](https://github.com/facebook/mysql-5.6/commit/7ddbc3c6e127cac00e40b035157b41e19da85638)

5. Null cherry-picked upstream commits:
- [1e138590f0bd791d32de003602294a7da31d9597 Added compression option --compress-data=<compressed_file_size_in_MB> to mysqldump program.](https://github.com/facebook/mysql-5.6/commit/1e138590f0bd791d32de003602294a7da31d9597)
- [5f38e203375f87ed86ac4ea0b2acc3121390e2c7 Add .watchmanconfig to mysql repo - this allows watchman to work properly when remoting from Atom](https://github.com/facebook/mysql-5.6/commit/5f38e203375f87ed86ac4ea0b2acc3121390e2c7)
- [27be29709113c9ba87b8e4f60ef52f0316e2a080 Fix a bug with whitespace linter](https://github.com/facebook/mysql-5.6/commit/27be29709113c9ba87b8e4f60ef52f0316e2a080)
- [151c05e272e76f3e04e02c9e6c725821c4f3ad66 Fix db_metadata not getting reset in db_load_opt if db.opt is of old format](https://github.com/facebook/mysql-5.6/commit/151c05e272e76f3e04e02c9e6c725821c4f3ad66)
- [6c47d0d2a6867b8924b30724ca655254b6327007 A new system variable to disable flow control between cordinator and worker threads](https://github.com/facebook/mysql-5.6/commit/6c47d0d2a6867b8924b30724ca655254b6327007)
- [c86625b4a61ec7880d50caa523cf0b5a809ab0d1 Add slave_gtid_info=optimized test to cover these scenarios](https://github.com/facebook/mysql-5.6/commit/c86625b4a61ec7880d50caa523cf0b5a809ab0d1)
- [5e3332972f5db09577f92743619c30f39fc18f1d Rename test to rpl_slave_flow_control_basic.test](https://github.com/facebook/mysql-5.6/commit/5e3332972f5db09577f92743619c30f39fc18f1d)
- [3b191ffa4aa28b295f711aa6e7605eb891ea4da1 Optimized implementation of integer to string converion](https://github.com/facebook/mysql-5.6/commit/3b191ffa4aa28b295f711aa6e7605eb891ea4da1)
- [a1d08a8dd274cdbbb3bcdcf3f5c17224bc40859c Statically link zstd with pic instead of using .so](https://github.com/facebook/mysql-5.6/commit/a1d08a8dd274cdbbb3bcdcf3f5c17224bc40859c)
- [67bd79355d1f284fa0bf9ce7a02e3f9a389e7718 Use xbstream instead of wdt by default](https://github.com/facebook/mysql-5.6/commit/67bd79355d1f284fa0bf9ce7a02e3f9a389e7718)
- [76b4fc3a95c50386dda97efc53eda93972f5ba1f Remove MySQL whitespace linter as it is not very precise when catching whitespace issues and we already have Clang](https://github.com/facebook/mysql-5.6/commit/76b4fc3a95c50386dda97efc53eda93972f5ba1f)
- [84de9ff2e63c827d1f89f688b251eefd37a6e276 Log SBR unsafe query to slow log](https://github.com/facebook/mysql-5.6/commit/84de9ff2e63c827d1f89f688b251eefd37a6e276)
- [96308865f3cde2feb0471902eb576fd9b785d96f Add a new option to make MySQL server truncate binlogs during crash recovery.](https://github.com/facebook/mysql-5.6/commit/96308865f3cde2feb0471902eb576fd9b785d96f)
- [34a5e1a7b04e67ee882abb2aa1b1b8c21d3b235d Fix writing format_description_event with different log_event_types.](https://github.com/facebook/mysql-5.6/commit/34a5e1a7b04e67ee882abb2aa1b1b8c21d3b235d)
- [f2d8ebde31393c0d2c82f8a6aec9eb369f64b131 Exit dump thread on error when rpl_wait_for_semi_sync_ack is enabled](https://github.com/facebook/mysql-5.6/commit/f2d8ebde31393c0d2c82f8a6aec9eb369f64b131)

### CODE REVIEW

**Comment by inikep** [comment]
Tested with `--big-test` at https://ps57.cd.percona.com/job/percona-server-5.7-param/237/

**Comment by bratao** [comment]
Thank you for doing the wonderful work of catching up with upstream MyRocks ❤️

File: mysql-test/include/restart_mysqld_with_option.inc [hunk 1]
```diff
@@ -0,0 +1,31 @@
+
```
Comment (george-lorch) [inline mysql-test/include/restart_mysqld_with_option.inc]: Why do we need this?  This functionality is covered in restart_mysqld.inc and IIRC we already converted many tests to using the 'modern' MySQL idiom in 5.7.

File: mysql-test/suite/rocksdb/include/use_direct_io_option.inc [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Common test pattern for options that control direct i/o
+#
+# Required input:
+#   $io_option - name and assignment to enable on server command line
+
+--source include/have_direct_io.inc
+
+--echo Checking direct reads
+--let $_mysqld_option=$io_option
+--source include/restart_mysqld_with_option.inc
```
Comment (george-lorch) [inline mysql-test/suite/rocksdb/include/use_direct_io_option.inc]: Use restart_mysqld.inc with restart_parameters instead?

File: storage/rocksdb/ha_rocksdb.cc [hunk 3]
```diff
@@ -106,6 +106,7 @@ static st_io_stall_stats io_stall_stats;
 const std::string DEFAULT_CF_NAME("default");
 const std::string DEFAULT_SYSTEM_CF_NAME("__system__");
 const std::string PER_INDEX_CF_NAME("$per_index_cf");
+
```
Comment (george-lorch) [inline storage/rocksdb/ha_rocksdb.cc:109]: Spurious change?

**Review by george-lorch — COMMENTED** [review]
Nice work, this was a tough one.  I did not do too much of a deep review but did make a few dives in more interesting places.  Looks good with minor comments.

File: mysql-test/include/restart_mysqld_with_option.inc (hunk 1 above)
Comment (inikep) [reply mysql-test/include/restart_mysqld_with_option.inc]: fixed

File: mysql-test/suite/rocksdb/include/use_direct_io_option.inc (hunk 2 above)
Comment (inikep) [reply mysql-test/suite/rocksdb/include/use_direct_io_option.inc]: fixed

File: storage/rocksdb/ha_rocksdb.cc (hunk 3 above)
Comment (inikep) [reply storage/rocksdb/ha_rocksdb.cc:109]: This is an upstream change so I prefer to keep it to minimize a diff with upstream.

Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc:109]: ack

**Review by george-lorch — APPROVED** [review]

**Pull request merged**